config/routes.rb

```diff
@@ -493,6 +495,60 @@
   lookup_user
 ].freeze
 
+def redirect_old_crud_actions(
+  old_controller: "",
+  new_controller: old_controller&.pluralize,
+  model: new_controller&.singularize,
+  actions: [:create, :destroy, :edit, :index, :show]
+)
+  return if old_controller.blank?
+
+  actions.each do |action|
+    case action
+    when :index # 3 old paths redirect to index
+      get "/#{old_controller}" =>
+        redirect("/#{new_controller}")
+      get "/#{old_controller}/index_#{model}" =>
+        redirect("/#{new_controller}")
+      get "/#{old_controller}/list_#{model}s" =>
+        redirect("/#{new_controller}")
+    when :show
+      get "/#{old_controller}/show_#{model}/:id" =>
+        redirect("/#{new_controller}/%{id}")
+    when :create
+      match "/#{old_controller}/create_#{model}",
+            to: redirect("/#{new_controller}/new"),
+            via: [:get, :post] # new, create
+    when :edit
+      match "/#{old_controller}/edit_#{model}/:id",
+            to: redirect("/#{new_controller}/%{id}/edit"),
+            via: [:get, :post] # edit, update
+    when :destroy
+      match "/#{old_controller}/destroy_#{model}",
+            to: redirect("#{new_controller}/%{id}"),
+            via: [:patch, :post, :put]
+    end
```

Comment:
Are you familiar with Martin Fowler’s book “Refactoring”?  He’s pretty much the de facto authority on refactoring.  He has this to say about switch/case statements in the current edition:

“ Repeated Switches

Talk to a true object-oriented evangelist and they’ll soon get onto the evils of switch statements. They’ll argue that any switch statement you see is begging for Replace Conditional with Polymorphism (272). We’ve even heard some people argue that all conditional logic should be replaced with polymorphism, tossing most ifs into the dustbin of history. Even in our more wild-eyed youth, we were never unconditionally opposed to the conditional. Indeed, the first edition of this book had a smell entitled “switch statements.” The smell was there because in the late 90’s we found polymorphism sadly underappreciated, and saw benefit in getting people to switch over. These days there is more polymorphism about, and it isn’t the simple red flag that it often was fifteen years ago. Furthermore, many languages support more sophisticated forms of switch statements that use more than some primitive code as their base. So we now focus on the repeated switch, where the same conditional switching logic (either in a switch/case statement or in a cascade of if/else statements) pops up in different places. The problem with such duplicate switches is that, whenever you add a clause, you have to find all the switches and update them. Against the dark forces of such repetition, polymorphism provides an elegant weapon for a more civilized codebase.”

However, in this case using polymorphism seems a bit crazy.  If there was more symmetry between the cases such that data was the only difference I’d consider using a hash, but I don’t see a simple way to do that so you might be stuck with a case statement.

Reply:
Looking at this more closely, I think this could be done with a hash.  That maps an action to a function that takes new_controller, old_controller, model, and id.  The code of each `when` are the different functions that go in the hash.

Reply:
I'll try doing this with a hash.
(I own Refactoring Ruby Edition (co-authored by Fowler) and consulted it when I ran into this problem. Same result: it also suggests polymorphism, which also seemed extreme to me (and I wasn't sure how to do it.)

Reply:
I don't understand what the hash key would do. I should be able to define a method (`redirect_<action>`) for each action, then iterate over the existing hash of actions.